django/thunderstore/repository/api/experimental/views.py

```diff
@@ -60,3 +68,31 @@ def get_no_cache_response(cls):
 
     def get_queryset(self):
         return get_mod_list_queryset(self.request.community_site)
+
+
+class UploadPackageApiView(APIView):
+    """
+    Uploads a package. Requires multipart/form-data.
+    """
+
+    parser_classes = [MultiPartParser]
+
+    def post(self, request):
+        categories = request.data.get("categories")
+        if categories:
+            try:
+                request.data["categories"] = json.loads(categories)
```

Comment:
This is used as multipart doesn't support arrays. Discord does a similar thing with [create message](https://discord.com/developers/docs/resources/channel#create-message), where it has a `payload_json` field which allows adding extra data that requires JSON (due to multipart's limitations).

Reply:
This should probably be moved to a field so the code's nicer and can be used elsewhere if we add other multipart endpoints

Reply:
This feels like it'll be a bit difficult for clients to encode properly. Ideally we'd have a json blob in the body / first part, and a file as an attachment / second part. Is there something preventing this in practice? What's going on here exactly, why do we need to manually decode this partially?

Reply:
It looks like this is a very common issue. Your idea of uploading a second (JSON) file is probably the best way, especially when we can most importantly standardise it across the API.

We can essentially write our own parser which will find a file with a specific name e.g. `json_payload` which will be the same across the entire API. We can also require this file to be `application/json` to be clear. With the parser, we can then unwrap this file and add it to `request.data`.

Keeping the field name the same and making it so *no* other fields exist other than actual files and `json_payload` should make it easier for users to understand and clients as instead of specific fields needing to be part of an extra JSON file and others being multipart fields, everything is in one place.

For example, for the package upload endpoint, we have the actual file field as `file` (the name of the multipart field). The client will also be required to upload another file with the field name `json_payload` which might look something like `{"author_name": "...", "has_nsfw_content": "..."}`. These fields will then be added to `request.data`, so it ends up looking something like:
```python
request.data = {
  "file": mod_zip,
  "author_name": "...",
  "has_nsfw_content": "...",
}
```
I think this would be the best way but I will also need to make sure lists aren't messed up as that was an issue I originally had (which is why JSON was required in the first place). The reason for unwrapping at the parser level is so passing into the serializer and form is easy. e.g.
```python
# View
serializer = PackageUploadSerializer(
    data=request.data,
    context={"request": request},
)

# Serializer
def _create_form(self, data) -> PackageUploadForm:
    request = self.context["request"]
    return PackageUploadForm(
        request.user,
        data.get("author_name"),
        request.community,
        data={
            "categories": data.get("categories"),
            "has_nsfw_content": data.get("has_nsfw_content"),
        },
        files={"file": data.get("file")},
    )
```
This serializer method could be simplified so the form's `data` argument takes all `data` fields other than `file` and `author_name`.

Reply:
What do you think of the field name `json_payload`? It's based on (me misremembering) Discord's [create message endpoint](https://discord.com/developers/docs/resources/channel#create-message) which uses `payload_json` but I haven't given it much thought.

Reply:
Actually I think forcing `json_payload` to have `application/json` and even be a file wouldn't add much but make it slightly more annoying for clients and make it harder to use tools like Postman.